mount/temp.go

```diff
@@ -59,6 +59,24 @@ func WithTempMount(ctx context.Context, mounts []Mount, f func(root string) erro
 			}
 		}
 	}()
+
+	// The volatile option of overlayfs doesn't allow to mount again using the
+	// same upper / work dirs. Since it's a temp mount, avoid using that
+	// option here.
+	// TODO: Make this logic conditional once the kernel supports reusing
+	// overlayfs volatile mounts.
+	for _, m := range mounts {
```

Comment:
IMO, we can remove this part. The caller should validate the `volatile` option at the beginning. 

In overlayFS snapshotter, we can check `volatile` in `NewSnapshotter` like https://github.com/containerd/containerd/blob/master/snapshots/overlay/overlayutils/check.go#L110 if set `Volatile=true` in toml.

Reply:
@fuweid Hi, I'm a co-worker of Mauricio.

Not sure I follow what you mean. If the kernel doesn't support this option, it will error out when trying to mount with volatile already. Do you suggest to move this code to a function and we just call it from here?

The code here makes sure to not use volatile for temp mounts, because the overlayfs dirs (workdir at least) is reused after these tmp mounts. But re-using them is not supported by the kernel when you use volatile, so we need to remove the option if not the following mounts that re-use the work-dir will fail.

Reply:
Hi @rata sorry for late reply. 

> => If the kernel doesn't support this option, it will error out when trying to mount with volatile already. 

I was thinking we can add pre-check option during initializing overlayFS snapshotter, like

```go
//  From https://github.com/containerd/containerd/blob/master/snapshots/overlay/overlay.go
//
// NewSnapshotter returns a Snapshotter which uses overlayfs. The overlayfs
// diffs are stored under the provided root. A metadata file is stored under
// the root.
func NewSnapshotter(root string, opts ...Opt) (snapshots.Snapshotter, error) {

     ...

     // figure out whether "index=off" option is recognized by the kernel
     var indexOff bool
     if _, err = os.Stat("/sys/module/overlay/parameters/index"); err == nil {
	indexOff = true
     }

     // check volatile option
     if config.volatile {
         if err := overlayutils.SupportVolatile(root); err != nil {
               return nil, err
         }
     }
    
     ...
}
``` 

The pre-check can help us to know the option is supported before running container.

> => But re-using them is not supported by the kernel when you use volatile

Yeah, you are correct. This code in `WithTempMount` should be kept. In addition, the other case is that container engine uses WithTempMount to get the user id mapping. I think use `volatile` is good for this case. But I don't have an idea how to use it correctly. 

-----------------------

The change looks good to me. I am fine with handling `overlayutils.SupportVolatile` in the follow-up pr.

LGTM

Reply:
@fuweid heh, no late response, it was quite fast :)

Ohh, I see. Yes, if we can merge this (one LGTM is missing IIUC) and I'll do a follow-up PR adding that it would be great.

Thanks again!

Reply:
btw, I think we might need to probe a dummy mount with the volatile option, not possible to do as we do with `indexOff` because that is a module _parameter_. Those are exposed on sysfs, not sure about mount options (think they are not). Will check and open a new PR in the next days with that :)

Reply:
@rata I would like to request @dmcgowan review on this~ Thanks!